File: 2020-02-Spring/0_arrays_matrix/word_search/python/Solution.py
def exist(board, word):
    """
    :type board: List[List[str]]
    :type word:  str
    :rtype: bool
    """
    rows = len(board)
    cols = len(board[0])
    for row in range(rows):
        for col in range(cols):
            if backtrack(row, col, word, board):
                return True
    return False

def backtrack(row, col, word, board):
    if (len(word) == 0):
        return True
    if row < 0 or row == len(board) or col < 0 or col == len(board[0]):
        return False
    if board[row][col] != word[0]:
        return False
    
    result = False
    board[row][col] = '*'
    for row_delta, col_delta in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
        result = backtrack(row + row_delta, col + col_delta, word[1:], board)
        if result: break
    
    board[row][col] = word[0]
    return result

def main():
    test_board = [['A','B','C','E'],
                  ['S','F','C','S'],
                  ['A','D','E','E' ]]
    print(exist(test_board, "ABCCED"))
    print(exist(test_board, "SEE"))
    print(exist(test_board, "ABCB"))

if __name__ == "__main__":
    main()

File: 2020-02-Spring/0_arrays_matrix/word_search/java/Solution.java
class Solution {
    
    public static void main(String[] args) {
        char[][] testBoard = {{ 'A','B','C','E' }, { 'S','F','C','S' }, { 'A','D','E','E' }};
        System.out.println(exist(testBoard, "ABCCED")); // Should print true
        System.out.println(exist(testBoard,"SEE"));     // Should print true
        System.out.println(exist(testBoard, "ABCB"));   // Should print false
    }
    public static boolean exist(char[][] board, String word) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (backtrack(board, 0, i, j, word)) {
                    return true;
                }
            }
        }
        return false;
    }

    private static boolean backtrack(char[][] board, int index, int row, int col, String word) {
        if (index == word.length()) {
            return true;
        }
        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || word.charAt(index) != board[row][col]) {
            return false;
        } else {
            board[row][col] = '*';
            boolean result = backtrack(board, index + 1, row + 1, col, word) ||
                             backtrack(board, index + 1, row, col + 1, word) ||
                             backtrack(board, index + 1, row - 1, col, word) ||
                             backtrack(board, index + 1, row, col - 1, word);
            board[row][col] = word.charAt(index);
            return result;
        }
    }
}